### PR TITLE
Expands capabilities of the scenario_set forsys call - 

### PR DESCRIPTION
**Pull request by strangeties**
Given a project area and priority conditions, stands are generated based on huc-12 boundaries and ranked by forsys according to condition rasters.

optional reviewers:
@elsiel23 
@JanLauGe 

This seems to work locally (e.g. http://localhost:8000/planscape-backend/forsys/scenario_set/)
image

@JanLauGe A raster is spit out at `condition_rasters = fetch_condition_rasters(priorities, region, project_area_raster)`. I still need to apply a mask (i.e. set values outside the projoct area to nan), but steps up to that point should be very similar between the patchmax and scenario set forsys calls - hopefully this can help with patchmax planning.

### CODE REVIEW

File: src/planscape/forsys/views.py [hunk 1]
```diff
@@ -3,43 +3,321 @@
...
-
 # Converts R dataframe to Pandas dataframe.
 # TODO: the broadly-accepted solution involves robjects.conversion.rpy2py - debug why it failed with an input type error.
 def convert_rdf_to_pddf(rdf):
   pddf = pd.DataFrame.from_dict({ key : np.asarray(rdf.rx2(key)) for key in rdf.names })
   return pddf
 
+# Converts dictionary of lists to R dataframe.
+# The lists must have equal length.
+def convert_dictionary_of_lists_to_rdf(lists):
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:28]: Have you thought about using a pandas Data Frame instead?

File: src/planscape/forsys/views.py [hunk 2]
```diff
@@ -3,43 +3,321 @@
...
+      data[key] = rpy2.robjects.StrVector(lists[key])
+    elif isinstance(el, float):
+      data[key] = rpy2.robjects.FloatVector(lists[key])
+    elif isinstance(el, int):
+      data[key] = rpy2.robjects.IntVector(lists[key])
+
+  rdf = rpy2.robjects.vectors.DataFrame(data)
+  return rdf
+
+# Translates polygon.
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:44]: translates from what to what? Looks to me like the input is a polygon and the output also is a polygon, but offset by some amount. Why is this needed?

File: src/planscape/forsys/views.py [hunk 3]
```diff
@@ -3,43 +3,321 @@
...
+    r = rasters[c]
+    d = r.bands[0].data()
+    count = np.count_nonzero(~np.isnan(d))
+    mean = np.sum(d[~np.isnan(d)]) / count
+    shape = np.shape(d)
+    raster_response.append("%s (non-nan area: %d, mean: %f, shape: %d x %d)"%(c, count, mean, shape[0], shape[1]))
+  return raster_response
+
+def get_condition_rasters(condition, region):
+  conditions = BaseCondition.objects.filter(condition_name=condition).filter(region_name=region).all()
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:72]: nit: If you expect exactly one condition, should this variable not be named using singular `condition` instead of `conditions`?

File: src/planscape/forsys/views.py [hunk 4]
```diff
@@ -3,43 +3,321 @@
...
+def raster_extent_overlaps_project_area(raster, project_area):
+  e = raster.extent
+  e_polygon = Polygon( ((e[0], e[1]),
+                        (e[2], e[1]),
+                        (e[2], e[3]),
+                        (e[0], e[3]),
+                        (e[0], e[1])) )
+  return e_polygon.overlaps(project_area)
+
+# Merges add-on raster with base_raster.
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:96]: Please expand this comment, I still don't know what this function is for. What are add-on raster and base_raster?

File: src/planscape/forsys/views.py [hunk 5]
```diff
@@ -3,43 +3,321 @@
...
+    # TODO: adjust raster data.
+    geo = project_area.intersection(b.geometry)
+    geo = translate_polygon(geo, 80, 0)
+    geo.transform(CoordTransform(SpatialReference(4269), SpatialReference(9822)))
+
+    # TODO: set project_area ID from an external source
+    data['proj_id'].append(1)
+    # TODO: double-check that it makes sense to use this ID.
+    data['stand_id'].append(b.id)
+    # TODO: double-check that this field is necessary.
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:225]: From the ForSys side we don't need it. We have `proj_id` and multi-part polygons are not a problem because the data is converted to table format at that stage already.

File: src/planscape/forsys/views.py [hunk 6]
```diff
@@ -3,43 +3,321 @@
...
+    data['shape_name'].append(b.shape_name)
+    data['area'].append(geo.area)
+    # TODO: adjust cost as a function of treatment type.
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:228]: So far I have operated under the assumption that users would just default to the cheapest treatment type. That is how I converted the multiple cost layers from John Hogland into one single cost layer. I think Min Su is investigating what we want to do going forward, we should check this with him.

File: src/planscape/forsys/views.py [hunk 7]
```diff
@@ -3,43 +3,321 @@
...
+
+# Runs a forsys scenario sets call.
+def run_forsys_scenario_sets(npdf, priorities):
+  kPriorityPrefix = "p"
+
+  import rpy2.robjects as robjects
+  robjects.r.source(os.path.join(settings.BASE_DIR, 'forsys/scenario_sets.R'))
+  scenario_sets_function_r = robjects.globalenv['scenario_sets']
+ 
+  # TODO: add inputs for thresholds.
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:247]: I suggest to use a dictionary for this, which should translate to a named list in R. Something like below can be used to generate the threshold string that ForSys requires. Note that only `>=` thresholds are supported right now.

```
{
  "variable_name_1": threshold_value_1,
  "variable_name_2": threshold_value_2
}

File: src/planscape/forsys/views.py [hunk 8]
```diff
@@ -3,43 +3,321 @@
...
+  # TODO: clean-up: pass header names (e.g. proj_id) into scenario_sets_function_r.
+  rdf = convert_dictionary_of_lists_to_rdf(npdf)
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:249]: again, I think a pandas DF would be the more intuitive input here

File: src/planscape/forsys/views.py [hunk 9]
```diff
@@ -3,43 +3,321 @@
...
+
+  priority_headers = []
+  for p in priorities:
+    priority_headers.append(kPriorityPrefix + "_" + p)
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:253]: This is confusing because you hard-coded `kPriorityPrefix = "p"` but here the variable `p` contains the string with the variable name. Since you're hard-coding `"p"`, why not change it to something like `priority_headers.append("p_" + p)` and make this a one-liner?

File: src/planscape/forsys/views.py [hunk 10]
```diff
@@ -3,43 +3,321 @@
...
+
+  forsys_output = scenario_sets_function_r(rdf,
+                                           robjects.FloatVector([np.ceil(np.sum(npdf['area']))]),
+                                           robjects.StrVector(priority_headers))
+
+  # TODO: add logic for applying constraints to forsys_output.
+
+  result = {}
+  result['top_stands'] = convert_rdf_to_pddf(forsys_output[0])
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:262]: ah, I see you're using pandas here 😃

File: src/planscape/forsys/views.py [hunk 11]
```diff
@@ -3,43 +3,321 @@
...
+                             (-120.18409937110482, 36.9321584213366),
+                             (-119.93422142411087, 36.94003252840713),
+                             (-120.03710286062301, 36.99713288358574),
+                             (-120.14015536869722, 37.05413814388948)) )
+    project_area.srid = 4269
+    if not project_area.valid:
+      raise ValueError("invalid project area: %s"%project_area.valid_reason) 
 
-    analysis = dummy_function_r(target_value)
+    # raster boundary seems to be 80 units off from the huc-12 boundaries.
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:285]: I don't understand this. Are you saying there is an issue with the RRK?

File: src/planscape/forsys/views.py [hunk 12]
```diff
@@ -3,43 +3,321 @@
...
-    for i in range(len(analysis)):
-      pddata = convert_rdf_to_pddf(analysis[i])
-      response['type_%d'%(i)] = pddata.to_json()
-    return JsonResponse(response)
+    if (save_debug_info):
+      response['debug'] = {}
+
+    # Filters boundaries by boundary_id.
+    # TODO: add more stand options. For the existing solution, project areas drawn manually are divided into stands according to HUC-12 boundaries.
+    boundaries = BoundaryDetails.objects.filter(boundary_id=huc12_id).filter(geometry__intersects=project_area)
```
Comment (JanLauGe) [inline src/planscape/forsys/views.py:297]: if we use intersect here we will have stands that end up in multiple projects. IMO we should use centroid instead. That way each cell (stand) can belong to exactly one project only (for non-overlapping projects)

**Review by JanLauGe — COMMENTED** [review]
This is great, thank you! Please have a look at my comments. I think I need your explanation on a couple of things.

File: src/planscape/forsys/views.py (hunk 1 above)
Comment (strangeties) [reply src/planscape/forsys/views.py:28]: I did consider passing pandas df instead of a dictionary of numpy lists; however, converting pandas df to R df using robjects.conversion.rpy2py requires more eng investigation/effort and I didn't think it'd be the best use of time when we already have a way to convert a dictionary of numpy lists into an R df.

File: src/planscape/forsys/views.py (hunk 2 above)
Comment (strangeties) [reply src/planscape/forsys/views.py:44]: This seems like a language thing: in grade school geometry classes, we learned about translations, rotations, flips, stretches. In this setting, a translation is a "slide" or "offset" - we move an object around a coordinate system without distorting it.

The offset is a temporary hack for making sure an intersection exists between a huc-12 boundary box and a condition raster. I noticed some offset (perhaps not exactly 80 degrees) between the two sources --

sample condition raster bounding boxes
image

sample huc-12 boundary extents:
image

I'd strongly prefer to check in code with the translation since that'll give us something that can be demo'd until the actual source of the offset is identified and fixed.

File: src/planscape/forsys/views.py (hunk 3 above)
Comment (strangeties) [reply src/planscape/forsys/views.py:72]: Acknowledged.

I name it conditions with the "s" because it's a list of conditions. We expect only 1 condition to be present, but that's something that needs to be verified with length checks.

File: src/planscape/forsys/views.py (hunk 4 above)
Comment (strangeties) [reply src/planscape/forsys/views.py:96]: I'll add more comments.

This is for merging two rasters into a single raster.
Sometimes a project area may cover multiple DB raster chunks.
A single raster is easier to pass around and run computations on.

File: src/planscape/forsys/views.py (hunk 5 above)
Comment (strangeties) [reply src/planscape/forsys/views.py:225]: Do you mean R scripts exist for this?
There's a difference between an R script and an implementation in production.

File: src/planscape/forsys/views.py (hunk 6 above)
Comment (strangeties) [reply src/planscape/forsys/views.py:228]: sg!

File: src/planscape/forsys/views.py (hunk 7 above)
Comment (strangeties) [reply src/planscape/forsys/views.py:247]: sg! let's not include this change in this particular pull request - once this is checked in, everyone can start attacking the TODO's.

File: src/planscape/forsys/views.py (hunk 8 above)
Comment (strangeties) [reply src/planscape/forsys/views.py:249]: the input to the R forsys call must be a r dataframe.

File: src/planscape/forsys/views.py (hunk 9 above)
Comment (strangeties) [reply src/planscape/forsys/views.py:253]: In terms of eng style, it's preferable to use constants rather than direct values.
How the constants are set in this file could be refined.

File: src/planscape/forsys/views.py (hunk 12 above)
Comment (strangeties) [reply src/planscape/forsys/views.py:297]: Yes: I think we can have multiple stand options, including centroid, raster pixel, etc. Let's check this in so that we have a solution that can be further refined by everyone.

Downstream, a polygon representing the intersection between these HUC-12 boundaries and a project area is computed as the "stand". So we will not have stands that wind up in multiple project areas.

File: src/planscape/forsys/scenario_sets.R [hunk 13]
```diff
@@ -0,0 +1,22 @@
+scenario_sets <- function(input_stand_data, max_area, priorities) {
+  library(forsys)
```
[the remark this answers is not in the document]
Comment (strangeties) [reply src/planscape/forsys/scenario_sets.R]: done.

File: src/planscape/forsys/scenario_sets.R [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+  stand_dat <- test_forest %>% st_drop_geometry()
+  run_outputs <- forsys::run(
+    return_outputs = TRUE,
+    scenario_name = "test_scenario",
+    stand_data = input_stand_data,
+    stand_id_field = "stand_id",
+    proj_id_field = "proj_id",
+    stand_area_field = "area",
+    scenario_priorities = priorities,
+    scenario_weighting_values = c("0 5 1"),
```
[the remark this answers is not in the document]
Comment (strangeties) [reply src/planscape/forsys/scenario_sets.R]: done.

File: src/planscape/forsys/views.py (hunk 4 above)
Comment (strangeties) [reply src/planscape/forsys/views.py]: Done - I've added more comments.
I hope it's clearer. gdal libraries tend to refer to this operation with terms like "mosaic" and "merge".

**Comment by strangeties** [comment]
Thanks for reviewing, Laurens!
When there's a chance, can you please take another look?
I think it'd be great to check this skeleton in soon so we can start refining the TODO's : D

File: src/planscape/forsys/views.py (hunk 2 above)
Comment (JanLauGe) [reply src/planscape/forsys/views.py:44]: SG, thanks for explaining
